### PR TITLE
[PW-7021] - Add failure counter for pay by link attempts

### PR DESCRIPTION
**Pull request by candemiralp**
**Description**

Pay By Link link is valid until the fifth failed payment attempt. Keep the order open until the fifth failure then cancel it.

For this purpose, a counter was added for the `Authorisation success=false` webhook notifications.

**Tested scenarios**

- Pay by Link

### CODE REVIEW

File: Helper/Webhook/AuthorisationWebhookHandler.php [hunk 1]
```diff
@@ -184,6 +185,13 @@ private function handleFailedAuthorisation(Order $order): Order
             return $order;
         }
 
+        // If the payment method is PBL, use failure counter before cancelling the order
+        if ($order->getPayment()->getMethod() == AdyenPayByLinkConfigProvider::CODE) {
+            if ($this->checkPaybylinkCancellation($order, $notification) === false) {
```
Comment (michaelpaul) [inline Helper/Webhook/AuthorisationWebhookHandler.php:190]:
```suggestion
            if (!$this->canCancelPayByLinkOrder($order, $notification)) {
```

File: Helper/Webhook/AuthorisationWebhookHandler.php [hunk 2]
```diff
@@ -229,4 +237,45 @@ private function handleManualCapture(Order $order, Notification $notification, b
...
         return $order;
     }
+
+    /**
+     * @param Order $order
+     * @param Notification $notification
+     * @return bool
+     * @throws \Exception
+     */
+    private function checkPaybylinkCancellation(Order $order, Notification $notification)
```
Comment (michaelpaul) [inline Helper/Webhook/AuthorisationWebhookHandler.php:247]:
```suggestion
    private function canCancelPayByLinkOrder(Order $order, Notification $notification): bool
```

File: Helper/Webhook/AuthorisationWebhookHandler.php [hunk 3]
```diff
@@ -229,4 +237,45 @@ private function handleManualCapture(Order $order, Notification $notification, b
...
+                    AdyenPayByLinkConfigProvider::MAX_FAILURE_COUNT
+                )),
+                $this->adyenLogger->getOrderContext($order)
+            );
+
+            return false;
+        }
+        else {
+            // Order can be cancelled.
+            return true;
```
Comment (michaelpaul) [inline Helper/Webhook/AuthorisationWebhookHandler.php:278]: We could convert this else into an `if` above like:

```php
if ($payByLinkFailureCount >= AdyenPayByLinkConfigProvider::MAX_FAILURE_COUNT) {
    // Order can be cancelled.
    return true;
}
// ... if block logic goes here ...
```

**Review by michaelpaul — COMMENTED** [review]
Looking good, I just got some small suggestions.

**Review by Morerice — APPROVED** [review]

**Review by michaelpaul — APPROVED** [review]

**Review by peterojo — APPROVED** [review]

**Pull request merged**